File: src/bmsx_cpp/machine/devices/vdp/readback.cpp
#include "machine/devices/vdp/readback.h"

#include <algorithm>
#include <cstring>

namespace bmsx {
void VdpReadbackUnit::resetSurfaceRegistry() {
	for (u32 surfaceId = 0u; surfaceId < VDP_RD_SURFACE_COUNT; ++surfaceId) {
		m_readSurfaces[surfaceId].surfaceId = surfaceId;
		m_readSurfaces[surfaceId].registered = false;
		invalidateSurface(surfaceId);
	}
}

void VdpReadbackUnit::registerSurface(u32 surfaceId) {
	m_readSurfaces[surfaceId].surfaceId = surfaceId;
	m_readSurfaces[surfaceId].registered = true;
	invalidateSurface(surfaceId);
}

void VdpReadbackUnit::invalidateSurface(u32 surfaceId) {
	m_readCaches[surfaceId].width = 0u;
}

void VdpReadbackUnit::beginFrame() {
	m_readBudgetBytes = ReadbackBudgetBytes;
	m_readOverflow = false;
}

u32 VdpReadbackUnit::status() const {
	u32 value = 0u;
	if (m_readBudgetBytes >= 4u) {
		value |= VDP_RD_STATUS_READY;
	}
	if (m_readOverflow) {
		value |= VDP_RD_STATUS_OVERFLOW;
	}
	return value;
}

bool VdpReadbackUnit::resolveSurface(u32 requestedSurfaceId, u32 mode) {
	faultCode = VDP_FAULT_NONE;
	faultDetail = 0u;
	advanceReadPosition = false;
	word = 0u;
	if (mode != VDP_RD_MODE_RGBA8888) {
		faultCode = VDP_FAULT_RD_UNSUPPORTED_MODE;
		faultDetail = mode;
		return false;
	}
	if (requestedSurfaceId >= VDP_RD_SURFACE_COUNT) {
		faultCode = VDP_FAULT_RD_SURFACE;
		faultDetail = requestedSurfaceId;
		return false;
	}
	const ReadSurface& readSurface = m_readSurfaces[requestedSurfaceId];
	if (!readSurface.registered) {
		faultCode = VDP_FAULT_RD_SURFACE;
		faultDetail = requestedSurfaceId;
		return false;
	}
	resolvedSurfaceId = readSurface.surfaceId;
	return true;
}

bool VdpReadbackUnit::readPixel(const VdpSurfaceUploadSlot& surface, u32 x, u32 y) {
	if (x >= surface.surfaceWidth || y >= surface.surfaceHeight) {
		faultCode = VDP_FAULT_RD_OOB;
		faultDetail = x | (y << 16u);
		word = 0u;
		return false;
	}
	if (m_readBudgetBytes < 4u) {
		m_readOverflow = true;
		word = 0u;
		advanceReadPosition = false;
		return true;
	}
	const ReadCache& cache = getReadCache(resolvedSurfaceId, surface, x, y);
	const u32 localX = x - cache.x0;
	const size_t byteIndex = static_cast<size_t>(localX) * 4u;
	const u32 r = cache.data[byteIndex];
	const u32 g = cache.data[byteIndex + 1u];
	const u32 b = cache.data[byteIndex + 2u];
	const u32 a = cache.data[byteIndex + 3u];
	m_readBudgetBytes -= 4u;
	nextX = x + 1u;
	nextY = y;
	if (nextX >= surface.surfaceWidth) {
		nextX = 0u;
		nextY = y + 1u;
	}
	advanceReadPosition = true;
	word = r | (g << 8u) | (b << 16u) | (a << 24u);
	return true;
}

VdpReadbackState VdpReadbackUnit::captureState() const {
	VdpReadbackState state;
	state.readBudgetBytes = m_readBudgetBytes;
	state.readOverflow = m_readOverflow;
	return state;
}

void VdpReadbackUnit::restoreState(const VdpReadbackState& state) {
	m_readBudgetBytes = state.readBudgetBytes;
	m_readOverflow = state.readOverflow;
}

VdpReadbackUnit::ReadCache& VdpReadbackUnit::getReadCache(u32 surfaceId, const VdpSurfaceUploadSlot& surface, u32 x, u32 y) {
	ReadCache& cache = m_readCaches[surfaceId];
	if (cache.width == 0u || cache.y != y || x < cache.x0 || x >= cache.x0 + cache.width) {
		prefetchReadCache(surfaceId, surface, x, y);
	}
	return cache;
}

// start numeric-sanitization-acceptable -- readback chunk width is the minimum of hardware cap, remaining surface span, and per-frame read budget.
void VdpReadbackUnit::prefetchReadCache(u32 surfaceId, const VdpSurfaceUploadSlot& surface, u32 x, u32 y) {
	const u32 maxPixelsByBudget = m_readBudgetBytes / 4u;
	if (maxPixelsByBudget == 0u) {
		m_readOverflow = true;
		m_readCaches[surfaceId].width = 0u;
		return;
	}
	const u32 chunkW = std::min(ReadbackMaxChunkPixels, std::min(surface.surfaceWidth - x, maxPixelsByBudget));
	ReadCache& cache = m_readCaches[surfaceId];
	copySurfacePixels(surface, x, y, chunkW, 1u, cache.data);
	cache.x0 = x;
	cache.y = y;
	cache.width = chunkW;
}
// end numeric-sanitization-acceptable

void VdpReadbackUnit::copySurfacePixels(const VdpSurfaceUploadSlot& surface, u32 x, u32 y, u32 width, u32 height, std::array<u8, ReadbackMaxChunkPixels * 4u>& out) {
	const u32 stride = surface.surfaceWidth * 4u;
	const u32 rowBytes = width * 4u;
	for (u32 row = 0u; row < height; ++row) {
		const size_t srcOffset = static_cast<size_t>(y + row) * static_cast<size_t>(stride) + static_cast<size_t>(x) * 4u;
		const size_t dstOffset = static_cast<size_t>(row) * static_cast<size_t>(rowBytes);
		std::memcpy(out.data() + dstOffset, surface.cpuReadback.data() + srcOffset, rowBytes);
	}
}

} // namespace bmsx

File: src/bmsx/machine/devices/vdp/readback.ts
import {
	VDP_FAULT_NONE,
	VDP_FAULT_RD_OOB,
	VDP_FAULT_RD_SURFACE,
	VDP_FAULT_RD_UNSUPPORTED_MODE,
	VDP_RD_MODE_RGBA8888,
	VDP_RD_STATUS_OVERFLOW,
	VDP_RD_STATUS_READY,
	VDP_RD_SURFACE_COUNT,
} from './contracts';
import type { VdpSurfaceUploadSlot } from './device_output';

export type VdpReadbackState = {
	readBudgetBytes: number;
	readOverflow: boolean;
};

type VdpReadSurface = {
	surfaceId: number;
	registered: boolean;
};

type VdpReadCache = {
	x0: number;
	y: number;
	width: number;
	data: Uint8Array;
};

const VDP_READBACK_BUDGET_BYTES = 4096;
const VDP_READBACK_MAX_CHUNK_PIXELS = 256;

function createReadSurfaceEntries(): VdpReadSurface[] {
	const entries: VdpReadSurface[] = [];
	for (let surfaceId = 0; surfaceId < VDP_RD_SURFACE_COUNT; surfaceId += 1) {
		entries.push({ surfaceId, registered: false });
	}
	return entries;
}

function createReadCaches(): VdpReadCache[] {
	const entries: VdpReadCache[] = [];
	for (let surfaceId = 0; surfaceId < VDP_RD_SURFACE_COUNT; surfaceId += 1) {
		entries.push({ x0: 0, y: 0, width: 0, data: new Uint8Array(VDP_READBACK_MAX_CHUNK_PIXELS * 4) });
	}
	return entries;
}

export class VdpReadbackUnit {
	public resolvedSurfaceId = 0;
	public faultCode = VDP_FAULT_NONE;
	public faultDetail = 0;
	public word = 0;
	public nextX = 0;
	public nextY = 0;
	public advanceReadPosition = false;
	private readonly readSurfaces = createReadSurfaceEntries();
	private readonly readCaches = createReadCaches();
	private readBudgetBytes = VDP_READBACK_BUDGET_BYTES;
	private readOverflow = false;

	public resetSurfaceRegistry(): void {
		for (let surfaceId = 0; surfaceId < VDP_RD_SURFACE_COUNT; surfaceId += 1) {
			const readSurface = this.readSurfaces[surfaceId]!;
			readSurface.surfaceId = surfaceId;
			readSurface.registered = false;
			this.invalidateSurface(surfaceId);
		}
	}

	public registerSurface(surfaceId: number): void {
		const readSurface = this.readSurfaces[surfaceId]!;
		readSurface.surfaceId = surfaceId;
		readSurface.registered = true;
		this.invalidateSurface(surfaceId);
	}

	public invalidateSurface(surfaceId: number): void {
		this.readCaches[surfaceId]!.width = 0;
	}

	public beginFrame(): void {
		this.readBudgetBytes = VDP_READBACK_BUDGET_BYTES;
		this.readOverflow = false;
	}

	public status(): number {
		let status = 0;
		if (this.readBudgetBytes >= 4) {
			status |= VDP_RD_STATUS_READY;
		}
		if (this.readOverflow) {
			status |= VDP_RD_STATUS_OVERFLOW;
		}
		return status;
	}

	public resolveSurface(requestedSurfaceId: number, mode: number): boolean {
		this.faultCode = VDP_FAULT_NONE;
		this.faultDetail = 0;
		this.advanceReadPosition = false;
		this.word = 0;
		if (mode !== VDP_RD_MODE_RGBA8888) {
			this.faultCode = VDP_FAULT_RD_UNSUPPORTED_MODE;
			this.faultDetail = mode;
			return false;
		}
		if (requestedSurfaceId >= VDP_RD_SURFACE_COUNT) {
			this.faultCode = VDP_FAULT_RD_SURFACE;
			this.faultDetail = requestedSurfaceId;
			return false;
		}
		const readSurface = this.readSurfaces[requestedSurfaceId]!;
		if (!readSurface.registered) {
			this.faultCode = VDP_FAULT_RD_SURFACE;
			this.faultDetail = requestedSurfaceId;
			return false;
		}
		this.resolvedSurfaceId = readSurface.surfaceId;
		return true;
	}

	public readPixel(surface: VdpSurfaceUploadSlot, x: number, y: number): boolean {
		const width = surface.surfaceWidth;
		const height = surface.surfaceHeight;
		if (x >= width || y >= height) {
			this.faultCode = VDP_FAULT_RD_OOB;
			this.faultDetail = (x | (y << 16)) >>> 0;
			this.word = 0;
			return false;
		}
		if (this.readBudgetBytes < 4) {
			this.readOverflow = true;
			this.word = 0;
			this.advanceReadPosition = false;
			return true;
		}
		const cache = this.getReadCache(this.resolvedSurfaceId, surface, x, y);
		const localX = x - cache.x0;
		const byteIndex = localX * 4;
		const r = cache.data[byteIndex]!;
		const g = cache.data[byteIndex + 1]!;
		const b = cache.data[byteIndex + 2]!;
		const a = cache.data[byteIndex + 3]!;
		this.readBudgetBytes -= 4;
		let nextX = x + 1;
		let nextY = y;
		if (nextX >= width) {
			nextX = 0;
			nextY = y + 1;
		}
		this.nextX = nextX;
		this.nextY = nextY;
		this.advanceReadPosition = true;
		this.word = (r | (g << 8) | (b << 16) | (a << 24)) >>> 0;
		return true;
	}

	public captureState(): VdpReadbackState {
		return {
			readBudgetBytes: this.readBudgetBytes,
			readOverflow: this.readOverflow,
		};
	}

	public restoreState(state: VdpReadbackState): void {
		this.readBudgetBytes = state.readBudgetBytes;
		this.readOverflow = state.readOverflow;
	}

	private getReadCache(surfaceId: number, surface: VdpSurfaceUploadSlot, x: number, y: number): VdpReadCache {
		const cache = this.readCaches[surfaceId]!;
		if (cache.width === 0 || cache.y !== y || x < cache.x0 || x >= cache.x0 + cache.width) {
			this.prefetchReadCache(cache, surface, x, y);
		}
		return cache;
	}

	private prefetchReadCache(cache: VdpReadCache, surface: VdpSurfaceUploadSlot, x: number, y: number): void {
		const maxPixelsByBudget = this.readBudgetBytes >>> 2;
		if (maxPixelsByBudget <= 0) {
			this.readOverflow = true;
			cache.width = 0;
			return;
		}
		const remainingWidth = surface.surfaceWidth - x;
		const chunkLimit = VDP_READBACK_MAX_CHUNK_PIXELS < remainingWidth ? VDP_READBACK_MAX_CHUNK_PIXELS : remainingWidth;
		const chunkW = chunkLimit < maxPixelsByBudget ? chunkLimit : maxPixelsByBudget;
		this.copySurfacePixels(cache, surface, x, y, chunkW, 1);
		cache.x0 = x;
		cache.y = y;
		cache.width = chunkW;
	}

	private copySurfacePixels(cache: VdpReadCache, surface: VdpSurfaceUploadSlot, x: number, y: number, width: number, height: number): void {
		const buffer = surface.cpuReadback;
		const stride = surface.surfaceWidth * 4;
		const rowBytes = width * 4;
		const out = cache.data;
		for (let row = 0; row < height; row += 1) {
			const srcOffset = (y + row) * stride + x * 4;
			const dstOffset = row * rowBytes;
			for (let byte = 0; byte < rowBytes; byte += 1) {
				out[dstOffset + byte] = buffer[srcOffset + byte]!;
			}
		}
	}
}
